accepted/2025/nuget-mcp.md

```diff
@@ -0,0 +1,265 @@
+
+## Explanation
+
+### Improve MCP server authoring
+
+#### Today's experience
+
+To create a .NET MCP server today, the experience is somewhat self-guided and results in a NuGet package that looks no different from any other .NET CLI tool.
+
+An MCP server author can create a .NET console app, consume the [ModelContextProtocol NuGet package](https://www.nuget.org/packages/ModelContextProtocol), and implement tools that will be available to the LLM. The SDK helps with the stdio-based protocol that allows the end user's IDE to launch, discover, and invoke tools.
+
+The author will decide what CLI arguments or environment variables are needed to invoke the MCP server.
+
+When the MCP server is ready, the author will pack the [project as a tool with `<PackAsTool>`](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create) and publish the NuGet package to NuGet.org (or their private package feed), so that it can be used by other people.
+
+This approach works, but has a couple of problems:
+
+- The package has no indication that it is an MCP server. It just has the `DotnetTool` package type like any other .NET tool.
+  - NuGet.org and other package sources have no way of filtering packages to "MCP" or tailoring the browsing experience.
+- The package provides no guidance to the end user on how to start the MCP server.
+  - The end user doesn't have a good way of knowing the needed `command`, `args`, and `env` settings they should use.
+- The package author needs to set up their project just right, with a console app, .NET tool, and MCP dependency.
+
+#### New package type
+
+To address the filtering and discoverability problem, we will introduce an additional NuGet package type of `McpServer`.
+
+Currently, .NET tools get the `DotnetTool` package type, defined in the resulting package manifest (.nuspec file) generated by `dotnet pack`. [As of .NET 10](https://github.com/dotnet/sdk/pull/48039), it is possible to add additional package types to .NET tool packages using the `<PackageType>McpServer</PackageType>` MSBuild property in the tool .csproj.
+
+The resulting MCP server .nupkg will have two package types: `DotnetTool` and `McpServer`. For more information on package types, [see NuGet documentation](https://learn.microsoft.com/en-us/nuget/reference/nuspec#packagetypes).
+
+#### Startup instructions
+
+We have two options to encode the startup instructions into the package:
+
+1. Instruct the package author to include the desired consumer MCP JSON in the readme, and allow NuGet.org to scrape the JSON from the README markdown (a code block matching a certain pattern).
+2. Embed standardized, machine-readable information in the package to allow tooling to know of the startup instructions required for the MCP server
```

Comment:
I prefer an explicit mcp.json in the package in case my README contains several examples of configuration where I show how to use certain environment variables etc.

Reply:
Yes, I agree. The risk with mcp.json (now renamed to [server.json apparently](https://github.com/modelcontextprotocol/registry/discussions/118#discussioncomment-13357690)) is that we ware buying into a specific schema as a contract. To me it feels like more of a commitment. If we scrape the README, it's a UI affordance only and not a contract with real package content consequences.

I would like more input on this area. I appreciate your comment!

Reply:
My preference would definitely be Option 2.  Given the schema is pretty ubiquitous, I don't see the commitment to it as being overly problematic.  If there's a new thing in the future, that could always be _additive_.

The advantage to having something more prescriptive is obviously easier detection and could enable more scenarios like auto-including these configs from referenced nuget packages in tooling to be contributed to the tool/ide itself.